crates/re_renderer/src/allocator/gpu_readback_belt.rs

```diff
@@ -234,54 +292,113 @@ impl GpuReadbackBelt {
+            chunk.last_received_frame_index + 1 < self.frame_index
+        });
+        self.received_chunks = retained;
+
+        for chunk in discarded {
+            re_log::debug!(
+                "Unread data from a GpuReadbackBelt was discarded. {} ranges remained unread.",
+                chunk.ranges_in_use.len()
+            );
+            self.reuse_chunk(chunk);
         }
     }
 
-    /// Receive all buffers that have been written by the GPU and are ready to be read by the CPU.
+    /// Try to receive a pending data readback with the given identifier and the given user data type.
+    ///
+    /// Returns the oldest received data with the given identifier & type.
+    /// This is *almost certainly* also the oldest scheduled readback. But there is no *strict* guarantee for this.
+    /// It could in theory happen that a readback is scheduled after a previous one, but finishes before it!
     ///
     /// ATTENTION: Do NOT assume any alignment on the slice passed to `on_data_received`.
     /// See this issue on [Alignment guarantees for mapped buffers](https://github.com/gfx-rs/wgpu/issues/3508).
-    ///
-    /// This should be called every frame to ensure that we're not accumulating too many buffers.
-    /// After this call, internal chunks will be re-used by subsequent [`Self::allocate`] calls.
-    pub fn receive_data(
+    pub fn readback_data<UserDataType: 'static>(
         &mut self,
-        mut on_data_received: impl FnMut(&[u8], GpuReadbackBufferIdentifier),
+        identifier: GpuReadbackIdentifier,
+        callback: impl FnOnce(&[u8], Box<UserDataType>),
     ) {
         crate::profile_function!();
 
-        while let Ok(mut chunk) = self.receiver.try_recv() {
-            {
-                let buffer_view = chunk
-                    .buffer
-                    .slice(..chunk.unused_offset())
-                    .get_mapped_range();
-
-                for (range, identifier) in chunk.ranges_in_use.drain(..) {
-                    on_data_received(
-                        &buffer_view[range.start as usize..range.end as usize],
-                        identifier,
-                    );
+        self.receive_chunks();
+
+        // Search for the user data in the readback chunks.
+        // A linear search is suited best since we expect the number both the number of pending chunks
+        // (typically just one or two!)
+        // as well as the number of readbacks per chunk to be small!
+        // Also note that identifiers may not be unique!
+        for (chunk_index, chunk) in self.received_chunks.iter_mut().enumerate() {
+            for (range_index, range) in chunk.ranges_in_use.iter().enumerate() {
+                if range.identifier != identifier || !range.user_data.is::<UserDataType>() {
+                    continue;
+                }
+
+                {
+                    let range = chunk.ranges_in_use.swap_remove(range_index);
```

Comment:
Wait, how is this legal? we're invalidating the iterator?!

Reply:
it's legal because there's a return after it. When writing this I was hoping so badly that it gets that and it did \o/

Reply:
but... there is no return? the only return that follows is conditional?!

Reply:
no, the return here is unconditional. It is after a conditional, but the return is not part of it

Reply:
or rather, the return is still conditional but it is in the same condition as the line you commented on. I.e. both swap_remove happen only when the loop is definitely exited